File: src/problems_all/35_search_insert_position/search_insert_position.java
/*
Runtime:    0 ms
Beats:      100.00%
Memory:     42.81 MB
Beats:      53.16%
Submission: https://leetcode.com/problems/search-insert-position/submissions/1439205025/
*/

class Solution {
    private int binarySearch(int[] arr, int lo, int hi, int x) {
        if (hi >= lo) {
            int mid = (hi + lo) / 2;

            if (mid >= arr.length) {
                return -1;
            }

            if (arr[mid] == x) {
                return mid;
            } else if (arr[mid] > x) {
                return binarySearch(arr, lo, mid - 1, x);
            } else {
                return binarySearch(arr, mid + 1, hi, x);
            }
        } else {
            return -1;
        }
    }

    public int searchInsert(int[] nums, int target) {
        int res = binarySearch(nums, 0, nums.length, target);
        if (res != -1) {
            return res;
        } else {
            for (int i = 0; i < nums.length; i++) {
                if (nums[i] > target) {
                    return i;
                }
            }

            return nums.length;
        }
    }

    public static void main(String[] args) {
        int[] nums = {1, 3, 5, 6};
        int target = 7;

        Solution solution = new Solution();
        int result = solution.searchInsert(nums, target);
        System.out.println(result);
    }
}

File: src/problems_all/35_search_insert_position/search_insert_position.py
"""
Runtime:    0 ms
Beats:      100.00%
Memory:     17.30 MB
Beats:      60.46%
Submission: https://leetcode.com/problems/search-insert-position/submissions/1439152588/
"""

from typing import List


class Solution:
    def binary_search(self, arr, lo, hi, x):
        if hi >= lo:
            mid = (hi + lo) // 2

            if arr[mid] == x:
                return mid

            elif arr[mid] > x:
                return self.binary_search(arr, lo, mid - 1, x)

            else:
                return self.binary_search(arr, mid + 1, hi, x)

        else:
            return -1

    def searchInsert(self, nums: List[int], target: int) -> int:
        res = self.binary_search(nums, 0, len(nums) - 1, target)
        if res != -1:
            return res

        else:
            for i, v in enumerate(nums):
                if v > target:
                    return i

            return len(nums)


if __name__ == '__main__':
    nums = [1, 3, 5, 6]
    target = 7
    result = Solution().searchInsert(nums, target)
    print(result)
